libtelemetry/src/full/java/com/mapbox/android/telemetry/UploadClientFactory.java

```diff
@@ -0,0 +1,88 @@
+package com.mapbox.android.telemetry;
+
+import android.content.Context;
+import android.content.pm.ApplicationInfo;
+import android.content.pm.PackageManager;
+import android.util.Log;
+
+import java.util.HashMap;
+import java.util.Map;
+
+class UploadClientFactory {
+  private static final String LOG_TAG = "TelemetryClientFactory";
+  private static final String RETRIEVING_APP_META_DATA_ERROR_MESSAGE = "Failed when retrieving app meta-data: %s";
+  private final String accessToken;
+  private final String userAgent;
+  private final CertificateBlacklist certificateBlacklist;
+  private final Context context;
+  private final Map<Environment, UploadClientBuild> BUILD_TELEMETRY_CLIENT = new HashMap<Environment,
```

Comment:
what's up with this map pattern again? :) can you please explain the benefit of it vs. simple `switch` in the static factory method?